File: digimon_gym/engine/data/scripts/bt21/bt21_074.py
from __future__ import annotations
from typing import TYPE_CHECKING, List, Dict, Any
from ....core.card_script import CardScript
from ....interfaces.card_effect import ICardEffect
from ....data.enums import EffectTiming

if TYPE_CHECKING:
    from ....core.card_source import CardSource


class BT21_074(CardScript):
    """BT21-074 Satellamon | Lv.5"""

    def get_card_effects(self, card: 'CardSource') -> List['ICardEffect']:
        effects = []

        # Factory effect: alt_digivolve_req
        # Alternate digivolution requirement
        effect0 = ICardEffect()
        effect0.set_effect_name("BT21-074 Alternate digivolution requirement")
        effect0.set_effect_description("Alternate digivolution requirement")
        # Alternate digivolution: alternate source for cost 3
        effect0._alt_digi_cost = 3

        def condition0(context: Dict[str, Any]) -> bool:
            permanent = card.permanent_of_this_card() if card else None
            if permanent and permanent.top_card:
                text = permanent.top_card.card_text
                if not ('Three Musketeers' in text):
                    return False
            else:
                return False
            return True
        effect0.set_can_use_condition(condition0)
        effects.append(effect0)

        # Factory effect: alt_digivolve_req
        # Alternate digivolution requirement
        effect1 = ICardEffect()
        effect1.set_effect_name("BT21-074 Alternate digivolution requirement")
        effect1.set_effect_description("Alternate digivolution requirement")
        # Alternate digivolution: with [Sup.] trait for cost 4
        effect1._alt_digi_cost = 3
        effect1._alt_digi_trait = "Sup."

        def condition1(context: Dict[str, Any]) -> bool:
            permanent = card.permanent_of_this_card() if card else None
            if not (permanent and permanent.top_card and (any('Sup.' in tr for tr in (getattr(permanent.top_card, 'card_traits', []) or [])))):
                return False
            return True
        effect1.set_can_use_condition(condition1)
        effects.append(effect1)

        # Timing: EffectTiming.OnEnterFieldAnyone
        # [On Play] By placing 1 [Appmon] or [Three Musketeers] trait card from your hand or trash as any of your Digimon's bottom digivolution cards, until your opponent's turn ends, their effects can't return that Digimon to hand or decks or affect it with <De-Digivolve> effects.
        effect2 = ICardEffect()
        effect2.set_timing(EffectTiming.OnEnterFieldAnyone)
        effect2.set_effect_name("BT21-074 Tuck to get protections")
        effect2.set_effect_description("[On Play] By placing 1 [Appmon] or [Three Musketeers] trait card from your hand or trash as any of your Digimon's bottom digivolution cards, until your opponent's turn ends, their effects can't return that Digimon to hand or decks or affect it with <De-Digivolve> effects.")
        effect2.is_on_play = True
        effect2._is_cannot_return_to_hand = True
        effect2._is_cannot_return_to_deck = True

        effect = effect2  # alias for condition closure
        def condition2(context: Dict[str, Any]) -> bool:
            # Triggered on play — validated by engine timing
            return True

        effect2.set_can_use_condition(condition2)

        def process2(ctx: Dict[str, Any]):
            """Action: Gain Keyword Cannot Return To Hand, Gain Keyword Cannot Return To Deck, Grant Bounce Immunity"""
            player = ctx.get('player')
            perm = ctx.get('permanent')
            game = ctx.get('game')
            if perm:
                perm.grant_keyword('_is_cannot_return_to_hand')
                perm.grant_keyword('_is_cannot_return_to_deck')
            # Prevent return to hand/deck via modifier system
            if perm and game:
                from digimon_gym.engine.interfaces.modifiers import ModifierType
                game.register_modifier(
                    ModifierType.CANNOT_BE_RETURNED, perm,
                    value_fn=lambda: True, expiry='end_of_turn')

        effect2.set_on_process_callback(process2)
        effects.append(effect2)

        # Timing: EffectTiming.OnEnterFieldAnyone
        # [When Digivolving] By placing 1 [Appmon] or [Three Musketeers] trait card from your hand or trash as any of your Digimon's bottom digivolution cards, until your opponent's turn ends, their effects can't return that Digimon to hand or decks or affect it with <De-Digivolve> effects.
        effect3 = ICardEffect()
        effect3.set_timing(EffectTiming.OnEnterFieldAnyone)
        effect3.set_effect_name("BT21-074 Tuck to get protections")
        effect3.set_effect_description("[When Digivolving] By placing 1 [Appmon] or [Three Musketeers] trait card from your hand or trash as any of your Digimon's bottom digivolution cards, until your opponent's turn ends, their effects can't return that Digimon to hand or decks or affect it with <De-Digivolve> effects.")
        effect3.is_when_digivolving = True
        effect3._is_cannot_return_to_hand = True
        effect3._is_cannot_return_to_deck = True

        effect = effect3  # alias for condition closure
        def condition3(context: Dict[str, Any]) -> bool:
            # Triggered when digivolving — validated by engine timing
            return True

        effect3.set_can_use_condition(condition3)

        def process3(ctx: Dict[str, Any]):
            """Action: Gain Keyword Cannot Return To Hand, Gain Keyword Cannot Return To Deck, Grant Bounce Immunity"""
            player = ctx.get('player')
            perm = ctx.get('permanent')
            game = ctx.get('game')
            if perm:
                perm.grant_keyword('_is_cannot_return_to_hand')
                perm.grant_keyword('_is_cannot_return_to_deck')
            # Prevent return to hand/deck via modifier system
            if perm and game:
                from digimon_gym.engine.interfaces.modifiers import ModifierType
                game.register_modifier(
                    ModifierType.CANNOT_BE_RETURNED, perm,
                    value_fn=lambda: True, expiry='end_of_turn')

        effect3.set_on_process_callback(process3)
        effects.append(effect3)

        # Timing: EffectTiming.OnEnterFieldAnyone
        # [When Digivolving][Once Per Turn] By trashing 1 card with the [Appmon] or [Three Musketeers] trait from your Digimon's digivolution cards, <De-Digivolve 1> 1 of your opponent's Digimon.
        effect4 = ICardEffect()
        effect4.set_timing(EffectTiming.OnEnterFieldAnyone)
        effect4.set_effect_name("BT21-074 Trash source to de-digivolve")
        effect4.set_effect_description("[When Digivolving][Once Per Turn] By trashing 1 card with the [Appmon] or [Three Musketeers] trait from your Digimon's digivolution cards, <De-Digivolve 1> 1 of your opponent's Digimon.")
        effect4.is_optional = True
        effect4.set_max_count_per_turn(1)
        effect4.set_hash_string("BT21_074De-digivolve")
        effect4.is_when_digivolving = True

        effect = effect4  # alias for condition closure
        def condition4(context: Dict[str, Any]) -> bool:
            # Triggered when digivolving — validated by engine timing
            return True

        effect4.set_can_use_condition(condition4)

        def process4(ctx: Dict[str, Any]):
            """Action: Trash Digivolution Cards, De Digivolve"""
            player = ctx.get('player')
            perm = ctx.get('permanent')
            game = ctx.get('game')
            # Trash digivolution cards from this permanent
            if perm and not perm.has_no_digivolution_cards:
                trashed = perm.trash_digivolution_cards(1)
                if player:
                    player.trash_cards.extend(trashed)
            if not (player and game):
                return
            def on_de_digivolve(target_perm):
                removed = target_perm.de_digivolve(1)
                enemy = player.enemy if player else None
                if enemy:
                    enemy.trash_cards.extend(removed)
            game.effect_select_opponent_permanent(
                player, on_de_digivolve, filter_fn=lambda p: p.is_digimon, is_optional=True)

        effect4.set_on_process_callback(process4)
        effects.append(effect4)

        # Timing: EffectTiming.OnAllyAttack
        # [When Digivolving][Once Per Turn] By trashing 1 card with the [Appmon] or [Three Musketeers] trait from your Digimon's digivolution cards, <De-Digivolve 1> 1 of your opponent's Digimon.
        effect5 = ICardEffect()
        effect5.set_timing(EffectTiming.OnAllyAttack)
        effect5.set_effect_name("BT21-074 Trash source to de-digivolve")
        effect5.set_effect_description("[When Digivolving][Once Per Turn] By trashing 1 card with the [Appmon] or [Three Musketeers] trait from your Digimon's digivolution cards, <De-Digivolve 1> 1 of your opponent's Digimon.")
        effect5.is_optional = True
        effect5.set_max_count_per_turn(1)
        effect5.set_hash_string("BT21_074De-digivolve")
        effect5.is_on_attack = True

        effect = effect5  # alias for condition closure
        def condition5(context: Dict[str, Any]) -> bool:
            # Triggered on attack — validated by engine timing
            return True

        effect5.set_can_use_condition(condition5)

        def process5(ctx: Dict[str, Any]):
            """Action: Trash Digivolution Cards, De Digivolve"""
            player = ctx.get('player')
            perm = ctx.get('permanent')
            game = ctx.get('game')
            # Trash digivolution cards from this permanent
            if perm and not perm.has_no_digivolution_cards:
                trashed = perm.trash_digivolution_cards(1)
                if player:
                    player.trash_cards.extend(trashed)
            if not (player and game):
                return
            def on_de_digivolve(target_perm):
                removed = target_perm.de_digivolve(1)
                enemy = player.enemy if player else None
                if enemy:
                    enemy.trash_cards.extend(removed)
            game.effect_select_opponent_permanent(
                player, on_de_digivolve, filter_fn=lambda p: p.is_digimon, is_optional=True)

        effect5.set_on_process_callback(process5)
        effects.append(effect5)

        # Timing: EffectTiming.WhenLinked
        # [When Linking] Delete 1 of your opponent's level 4 or lower Digimon.
        effect6 = ICardEffect()
        effect6.set_timing(EffectTiming.WhenLinked)
        effect6.set_effect_name("BT21-074 Delete level 4")
        effect6.set_effect_description("[When Linking] Delete 1 of your opponent's level 4 or lower Digimon.")

        effect = effect6  # alias for condition closure
        def condition6(context: Dict[str, Any]) -> bool:
            return True

        effect6.set_can_use_condition(condition6)

        def process6(ctx: Dict[str, Any]):
            """Action: Delete"""
            player = ctx.get('player')
            perm = ctx.get('permanent')
            game = ctx.get('game')
            if not (player and game):
                return
            def target_filter(p):
                if p.level is None or p.level > 4:
                    return False
                return p.is_digimon
            def on_delete(target_perm):
                enemy = player.enemy if player else None
                if enemy:
                    enemy.delete_permanent(target_perm)
            game.effect_select_opponent_permanent(
                player, on_delete, filter_fn=target_filter, is_optional=False)

        effect6.set_on_process_callback(process6)
        effects.append(effect6)

        return effects

File: code/digimon-engine/tests/cards_behavioral/bt21/bt21_074.rs
//! BT21-074 Satellamon — Digimon (Appmon), Lv.5, Purple/Black, DP 7000, Cost 7.
//! Traits: Ult. / Appmon / GPS. Attribute: Navi.
//!
//! # Card text (data/card_bundles/BT21-074.md — official Bandai DB, confirmed
//! against the card image BT21-074.webp, which also prints the "Sup. 4"
//! digivolve circle the DB bundle omits)
//!
//! Digivolve: Purple Lv.4 / cost 4; Black Lv.4 / cost 4; [Sup.] trait / cost 4
//! (image + DCGO); and [Digivolve] Lv.4 w/[Three Musketeers] in text: Cost 3.
//!
//! [On Play] [When Digivolving] By placing 1 [Appmon] or [Three Musketeers]
//! trait card from your hand or trash as any of your Digimon's bottom
//! digivolution card, until your opponent's turn ends, their effects can't
//! return that Digimon to hands or decks or affect it with ＜De-Digivolve＞
//! effects. [When Digivolving] [When Attacking] [Once Per Turn] By trashing
//! 1 card with the [Appmon] or [Three Musketeers] trait from your Digimon's
//! digivolution cards, ＜De-Digivolve 1＞ 1 of your opponent's Digimon.
//! Special Rule: +4000 DP (link box)
//! Inherited: ＜Link＞ [Appmon] trait: Cost 3  [When Linking] Delete 1 of your
//! opponent's level 4 or lower Digimon.
//!
//! # DCGO C# reference
//! DCGO/Assets/Scripts/CardEffect/BT21/Purple/BT21_074.cs
//! - Alt digivolution: Level == 4 && HasText("Three Musketeers") cost 3;
//!   EqualsTraits("Sup.") cost 4. Link condition: HasAppmonTraits, cost 3.
//! - On Play / When Digivolving (`-1, false` = mandatory clause): hand-or-
//!   trash bool select, then SelectHandEffect / SelectCardEffect(Root.Trash)
//!   (canNoSelect: true) over EqualsTraits("Appmon") || HasThreeMusketeersTraits;
//!   then SelectPermanentEffect over own non-token Digimon (canNoSelect: true)
//!   → AddDigivolutionCardsBottom; then GainCanNotReturnToHand /
//!   GainCanNotReturnToDeck (IsOpponentEffect, UntilOpponentTurnEnd) +
//!   ImmuneFromDeDigivolveClass (UntilOpponentTurnEndEffects).
//! - When Digivolving / When Attacking (`1, true` = OPT + optional, shared
//!   hash "BT21_074De-digivolve"): SelectTrashDigivolutionCards over any own
//!   Digimon (canNoTrash: false) filtered Appmon-or-TM trait; if trashed →
//!   mandatory SelectPermanentEffect over opponent Digimon → IDegeneration(1).
//! - WhenLinked: mandatory Destroy select over opponent Digimon with level ≤ 4.
//!
//! # Patterns this test covers (RUST_DSL_TEST_API §4.3)
//! - H alt paths (two circles + trait gate + in_text_contains special)
//! - A6 union-zone (hand|trash) place-as-bottom-source + opponent-scoped
//!   return/De-Digivolve protection modifiers with end_of_opponents_turn expiry
//! - E2 OPT optional [WD][WA] union-material trash cost → De-Digivolve 1
//! - DigiLink self link-condition + linked DP aura + when_linked delete

#![allow(dead_code, unused_imports, unused_variables, unused_mut)]

use digimon_dsl::compiled::{
    CompiledAltPathKind, CompiledClause, CompiledCost, CompiledDeclarativeClause, CompiledScope,
    CompiledStep, CompiledTiming,
};
use digimon_engine::action::space::{encode_attack, PASS, TRASH_EFFECT_START};
use digimon_engine::card_data::CardData;
use digimon_engine::debug_runner::{make_test_card, DebugRunner, DebugRunnerBuilder};
use digimon_engine::enums::CardColor;
use digimon_engine::enums::{CardKind, EffectTiming, ModifierType, PlayerId};
use digimon_engine::permanent::PermanentHandle;
use digimon_engine::selection::{SelectionKind, TriggerSource, UnionZoneSet};

const CARD_ID: &str = "BT21-074";

fn digimon(id: &str, level: u8, dp: i32, cost: u16, traits: &[&str]) -> CardData {
    let mut c = make_test_card(id, id);
    c.card_kind = CardKind::Digimon;
    c.level = Some(level);
    c.dp = Some(dp);
    c.play_cost = cost;
    c.colors = vec![CardColor::Purple];
    c.traits = traits.iter().map(|t| t.to_string()).collect();
    c
}

fn base() -> DebugRunnerBuilder {
    DebugRunner::builder()
        .dsl_card(CARD_ID)
        .expect("BT21-074 YAML parses and compiles")
        .add_card(make_test_card("DECK-PAD", "DECK-PAD"))
        .add_card(digimon("HOST-APP", 4, 4000, 4, &["Appmon"]))
        .add_card(digimon("HOST-PLAIN", 4, 4000, 4, &["Beast"]))
        .add_card(digimon("APP-CARD", 3, 2000, 3, &["Appmon"]))
        .add_card(digimon("TM-CARD", 3, 2000, 3, &["Three Musketeers"]))
        .add_card(digimon("PLAIN-CARD", 3, 2000, 3, &["Beast"]))
        .add_card(digimon("OPP-BASE", 3, 2000, 3, &["Beast"]))
        .add_card(digimon("OPP-TOP", 4, 4000, 4, &["Beast"]))
        .add_card(digimon("OPP-LV4", 4, 4000, 4, &["Beast"]))
        .add_card(digimon("OPP-LV5", 5, 6000, 6, &["Beast"]))
        .deck(0, &["DECK-PAD"; 6])
        .deck(1, &["DECK-PAD"; 6])
}

fn fire(runner: &mut DebugRunner, timing: EffectTiming, perm: PermanentHandle) {
    runner
        .game
        .enqueue_triggered(timing, TriggerSource::Permanent(perm));
    runner.game.drain_effect_queue();
}

fn resolve_trigger_order_if_present(runner: &mut DebugRunner) {
    while matches!(runner.pending_kind(), Some(SelectionKind::TriggerOrder)) {
        let act = runner.pending_selection().unwrap().valid_action_ids[0];
        runner.execute_action(0, act).expect("TriggerOrder");
    }
}

fn fire_link_onto_host(runner: &mut DebugRunner, host: PermanentHandle) {
    let linked = runner.push_linked_owned(host, CARD_ID, 0);
    for pid in 0..2u8 {
        runner.game.enqueue_triggered(
            EffectTiming::OnLink,
            TriggerSource::Linked {
                player: pid as PlayerId,
                host,
                card: linked,
            },
        );
    }
    runner.game.drain_effect_queue();
}

fn has_protection(runner: &DebugRunner, h: PermanentHandle) -> bool {
    runner
        .game
        .modifiers
        .has(h, ModifierType::CannotBeReturnedToHand)
        && runner
            .game
            .modifiers
            .has(h, ModifierType::CannotBeReturnedToDeck)
        && runner
            .game
            .modifiers
            .has(h, ModifierType::CannotBeDeDigivolved)
}

// ─── Section 1 — Structural assertions ───────────────────────────────────────

#[test]
fn bt21_074_yaml_has_printed_metadata() {
    let runner = base().start();
    let card = runner
        .compiled_card(CARD_ID)
        .expect("BT21-074 in embedded pack");
    assert_eq!(card.name, "Satellamon");
    assert_eq!(card.level, Some(5));
    assert_eq!(card.cost, Some(7));
    assert_eq!(card.dp, Some(7000));
    for t in ["Ult.", "Appmon", "GPS"] {
        assert!(card.traits.contains(&t.to_string()), "missing trait {t}");
    }
}

#[test]
fn bt21_074_alt_paths() {
    let runner = base().start();
    let card = runner.compiled_card(CARD_ID).expect("compiled");
    let circles = card
        .alt_paths
        .iter()
        .filter(|p| {
            p.kind == CompiledAltPathKind::Digivolve
                && p.cost == Some(CompiledCost::Literal(4))
                && p.from
                    .as_ref()
                    .is_some_and(|f| f.level_eq == Some(4) && f.color_is.is_some())
        })
        .count();
    assert_eq!(circles, 2, "Purple Lv.4 / 4 and Black Lv.4 / 4");
    assert!(
        card.alt_paths.iter().any(|p| {
            p.kind == CompiledAltPathKind::Digivolve
                && p.cost == Some(CompiledCost::Literal(4))
                && p.from
                    .as_ref()
                    .is_some_and(|f| f.trait_has.as_deref() == Some("Sup."))
        }),
        "[Sup.] trait / cost 4 (image + DCGO)"
    );
    assert!(
        card.alt_paths.iter().any(|p| {
            p.kind == CompiledAltPathKind::Digivolve
                && p.cost == Some(CompiledCost::Literal(3))
                && p.from.as_ref().is_some_and(|f| {
                    f.level_eq == Some(4)
                        && f.in_text_contains.as_deref() == Some("Three Musketeers")
                })
        }),
        "[Digivolve] Lv.4 w/[Three Musketeers] in text: Cost 3"
    );
}

#[test]
fn bt21_074_link_aura_and_clause_shapes() {
    let runner = base().start();
    let card = runner.compiled_card(CARD_ID).expect("compiled");
    assert!(card.effects.iter().any(|c| matches!(
        c,
        CompiledClause::Declarative(CompiledDeclarativeClause::LinkCondition { cost, .. }) if *cost == 3
    )), "<Link> [Appmon] trait: Cost 3");
    assert!(
        card.effects.iter().any(|c| matches!(
            c,
            CompiledClause::Declarative(CompiledDeclarativeClause::Aura { scope, dp_modifier, .. })
                if *scope == CompiledScope::Linked && *dp_modifier == Some(4000)
        )),
        "link box +4000 DP aura"
    );
    let triggered: Vec<_> = card
        .effects
        .iter()
        .filter_map(|c| match c {
            CompiledClause::Triggered(t) => Some(t),
            _ => None,
        })
        .collect();
    let tuck = triggered
        .iter()
        .find(|t| t.when.contains(&CompiledTiming::OnPlay))
        .expect("[On Play][When Digivolving] tuck clause");
    assert!(tuck.when.contains(&CompiledTiming::WhenDigivolving));
    assert!(
        !tuck.optional,
        "DCGO isOptional: false (the picks themselves are declinable)"
    );
    let dd = triggered
        .iter()
        .find(|t| t.when.contains(&CompiledTiming::WhenAttacking))
        .expect("[When Digivolving][When Attacking] De-Digivolve clause");
    assert!(dd.when.contains(&CompiledTiming::WhenDigivolving));
    assert!(dd.once_per_turn);
    assert!(dd.optional);
    let wl = triggered
        .iter()
        .find(|t| t.when == vec![CompiledTiming::WhenLinked])
        .expect("[When Linking]");
    assert_eq!(wl.scope, CompiledScope::Linked);
}

// ─── Section 2 — [On Play][WD] tuck from hand/trash → protection ─────────────

#[test]
fn bt21_074_on_play_tucks_from_hand_and_protects_chosen_digimon_until_opponents_turn_ends() {
    let mut runner = base()
        .hand(0, &["APP-CARD", "PLAIN-CARD"])
        .memory(5)
        .start();
    let sat = runner.place_on_field(0, CARD_ID, Some(0));
    let host = runner.place_on_field(0, "HOST-PLAIN", Some(0));

    fire(&mut runner, EffectTiming::OnPlay, sat);

    assert_eq!(
        runner.pending_kind(),
        Some(SelectionKind::UnionZone {
            zones: UnionZoneSet::HAND | UnionZoneSet::TRASH
        })
    );
    assert!(runner.pending_is_optional(), "DCGO canNoSelect: true");
    let view = runner.pending_selection_view().unwrap();
    assert_eq!(
        view.valid_action_ids.iter().filter(|&&a| a != PASS).count(),
        1,
        "only the [Appmon] card qualifies (PLAIN-CARD does not)"
    );
    runner
        .execute_action(0, view.valid_action_ids[0])
        .expect("pick APP-CARD");

    assert_eq!(runner.pending_kind(), Some(SelectionKind::OwnField));
    runner
        .execute_action(0, encode_attack(0, host.index as u16))
        .expect("tuck under HOST-PLAIN");
    runner.auto_resolve().ok();

    let stack = &runner.game.players[0].battle_area[host.index as usize].card_sources;
    assert_eq!(stack.len(), 2);
    assert_eq!(
        stack[0].card_id(&runner.game.card_data),
        "APP-CARD",
        "bottom source"
    );
    assert_eq!(runner.hand_size(0), 1);
    assert!(
        has_protection(&runner, host),
        "return-to-hand/deck + De-Digivolve protection"
    );
    assert!(
        !has_protection(&runner, sat),
        "only THAT Digimon is protected"
    );

    runner.game.end_turn(); // P0 → P1
    runner.auto_resolve().ok();
    assert!(
        has_protection(&runner, host),
        "still protected during the opponent's turn"
    );
    runner.game.end_turn(); // P1 → P0
    runner.auto_resolve().ok();
    assert!(
        !has_protection(&runner, host),
        "expires when the opponent's turn ends"
    );
}

#[test]
fn bt21_074_on_play_tucks_from_trash() {
    let mut runner = base().memory(5).start();
    runner.inject_trash(0, "TM-CARD");
    runner.inject_trash(0, "PLAIN-CARD");
    let sat = runner.place_on_field(0, CARD_ID, Some(0));
    let host = runner.place_on_field(0, "HOST-PLAIN", Some(0));

    fire(&mut runner, EffectTiming::OnPlay, sat);
    let view = runner.pending_selection_view().expect("union pick");
    let idx = runner.game.players[0]
        .trash
        .iter()
        .position(|c| c.card_id(&runner.game.card_data) == "TM-CARD")
        .unwrap();
    let pick = TRASH_EFFECT_START + idx as u16;
    assert!(view.valid_action_ids.contains(&pick));
    assert_eq!(
        view.valid_action_ids.iter().filter(|&&a| a != PASS).count(),
        1
    );
    runner
        .execute_action(0, pick)
        .expect("pick TM-CARD from trash");
    runner
        .execute_action(0, encode_attack(0, host.index as u16))
        .expect("tuck");
    runner.auto_resolve().ok();

    let stack = &runner.game.players[0].battle_area[host.index as usize].card_sources;
    assert_eq!(stack[0].card_id(&runner.game.card_data), "TM-CARD");
    assert!(has_protection(&runner, host));
}

#[test]
fn bt21_074_on_play_decline_tucks_nothing_and_protects_nothing() {
    let mut runner = base().hand(0, &["APP-CARD"]).memory(5).start();
    let sat = runner.place_on_field(0, CARD_ID, Some(0));
    let host = runner.place_on_field(0, "HOST-PLAIN", Some(0));
    fire(&mut runner, EffectTiming::OnPlay, sat);
    runner.execute_action(0, PASS).expect("decline");
    assert!(runner.pending_selection().is_none());
    assert_eq!(runner.hand_size(0), 1);
    assert!(!has_protection(&runner, host));
}

#[test]
fn bt21_074_on_play_no_eligible_card_no_prompt() {
    let mut runner = base().hand(0, &["PLAIN-CARD"]).memory(5).start();
    let sat = runner.place_on_field(0, CARD_ID, Some(0));
    runner.place_on_field(0, "HOST-PLAIN", Some(0));
    fire(&mut runner, EffectTiming::OnPlay, sat);
    assert!(
        runner.pending_selection().is_none(),
        "unexpected prompt: {:?}",
        runner.pending_selection_view()
    );
}

// ─── Section 3 — [WD][WA][OPT] trash Appmon/TM source → De-Digivolve 1 ──────

fn dd_runner() -> (DebugRunner, PermanentHandle, PermanentHandle) {
    let mut runner = base().memory(5).start();
    let sat = runner.place_on_field(0, CARD_ID, Some(0));
    let host = runner.place_on_field(0, "HOST-APP", Some(0));
    runner.push_source(host, "TM-CARD");
    runner.push_source(host, "PLAIN-CARD");
    let opp = runner.place_stack(1, &["OPP-BASE", "OPP-TOP"]);
    (runner, sat, opp)
}

#[test]
fn bt21_074_wa_trashes_source_then_de_digivolves() {
    let (mut runner, sat, opp) = dd_runner();
    fire(&mut runner, EffectTiming::WhenAttacking, sat);
    assert_eq!(runner.pending_kind(), Some(SelectionKind::Replacement));
    runner.accept_optional_trigger().expect("accept");
    assert_eq!(
        runner.pending_kind(),
        Some(SelectionKind::UnionZone {
            zones: UnionZoneSet::MATERIAL
        })
    );
    let view = runner.pending_selection_view().unwrap();
    assert_eq!(
        view.valid_action_ids.iter().filter(|&&a| a != PASS).count(),
        1,
        "only TM-CARD"
    );
    runner
        .execute_action(0, view.valid_action_ids[0])
        .expect("trash");
    assert_eq!(runner.pending_kind(), Some(SelectionKind::OppField));
    runner.auto_resolve().ok();
    assert_eq!(
        runner.game.players[1].battle_area[opp.index as usize]
            .top_card()
            .card_id(&runner.game.card_data),
        "OPP-BASE"
    );
}

#[test]
fn bt21_074_wa_decline_leaves_everything_untouched() {
    let (mut runner, sat, opp) = dd_runner();
    fire(&mut runner, EffectTiming::WhenAttacking, sat);
    runner.decline_optional_trigger().expect("decline");
    assert!(runner.pending_selection().is_none());
    assert_eq!(
        runner.game.players[1].battle_area[opp.index as usize]
            .card_sources
            .len(),
        2
    );
}

#[test]
fn bt21_074_wa_opt_locks_second_activation_same_turn() {
    let (mut runner, sat, opp) = dd_runner();
    runner.push_source(sat, "APP-CARD"); // second eligible source
    fire(&mut runner, EffectTiming::WhenAttacking, sat);
    runner.accept_optional_trigger().expect("accept");
    runner.auto_resolve().expect("first activation");
    fire(&mut runner, EffectTiming::WhenAttacking, sat);
    assert!(runner.pending_selection().is_none(), "OPT lock");
    runner.game.end_turn();
    runner.auto_resolve().ok();
    runner.game.end_turn();
    runner.auto_resolve().ok();
    fire(&mut runner, EffectTiming::WhenAttacking, sat);
    assert_eq!(
        runner.pending_kind(),
        Some(SelectionKind::Replacement),
        "lock cleared"
    );
}

/// On [When Digivolving] both clauses fire: the tuck clause AND the OPT
/// De-Digivolve clause share the timing (a TriggerOrder pick may surface).
#[test]
fn bt21_074_wd_fires_both_clauses() {
    let mut runner = base().hand(0, &["APP-CARD"]).memory(5).start();
    let sat = runner.place_on_field(0, CARD_ID, Some(0));
    let host = runner.place_on_field(0, "HOST-APP", Some(0));
    runner.push_source(host, "TM-CARD");
    let opp = runner.place_stack(1, &["OPP-BASE", "OPP-TOP"]);
    fire(&mut runner, EffectTiming::WhenDigivolving, sat);
    resolve_trigger_order_if_present(&mut runner);
    // Drive every prompt by accepting / taking the first legal action.
    while let Some(view) = runner.pending_selection_view() {
        let a = view
            .valid_action_ids
            .iter()
            .copied()
            .find(|&id| id != PASS)
            .unwrap_or(PASS);
        runner
            .execute_action(view.selecting_player, a)
            .expect("drive");
        resolve_trigger_order_if_present(&mut runner);
    }
    assert_eq!(runner.hand_size(0), 0, "APP-CARD was tucked");
    assert_eq!(
        runner.game.players[1].battle_area[opp.index as usize]
            .top_card()
            .card_id(&runner.game.card_data),
        "OPP-BASE",
        "the opponent was De-Digivolved"
    );
}

// ─── Section 4 — Link box + [When Linking] ────────────────────────────────────

#[test]
fn bt21_074_linked_dp_bonus_4000_reaches_host() {
    let mut runner = base().memory(5).start();
    let host = runner.place_on_field(0, "HOST-APP", Some(0));
    let dp_before = runner.game.effective_dp(host).unwrap();
    runner.push_linked_owned(host, CARD_ID, 0);
    runner.game.tick_declarative_effects();
    assert_eq!(runner.game.effective_dp(host), Some(dp_before + 4000));
}

#[test]
fn bt21_074_when_linked_deletes_opp_level_4_or_lower_only() {
    let mut runner = base().memory(5).start();
    let host = runner.place_on_field(0, "HOST-APP", Some(0));
    runner.place_on_field(1, "OPP-LV4", Some(0));
    runner.place_on_field(1, "OPP-LV5", Some(0));
    fire_link_onto_host(&mut runner, host);
    assert_eq!(runner.pending_kind(), Some(SelectionKind::OppField));
    let view = runner.pending_selection_view().unwrap();
    assert_eq!(
        view.valid_action_ids.iter().filter(|&&a| a != PASS).count(),
        1
    );
    runner.auto_resolve().ok();
    assert_eq!(runner.battle_area_size(1), 1);
    assert!(runner.game.players[1]
        .battle_area
        .iter()
        .all(|p| p.top_card().card_id(&runner.game.card_data) == "OPP-LV5"));
}
